### PR TITLE
OOP school library: decorate a class

### PR DESCRIPTION
**Pull request by habasm**
### Project requirements
- Check the usage of [Decorator pattern in Ruby](https://refactoring.guru/design-patterns/decorator/ruby/example#example-0)
- Think about how you can use two decorators in order to capitalize and trim people's names.
- In the following steps you will get instructions for implementing the Decorator design pattern for this case.

#### Interface
- Create a class `Nameable`.
   - Implement a method called `correct_name` that will raise a `NotImplementedError`.

#### Turn your Person class to Nameable
- Make sure that your Person class inherits from Nameable
- Make sure that this class has a method `correct_name` implemented. It should simply return the name attribute.

#### Prepare base Decorator
- Make sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implement the `correct_name` method that returns the result of the correct_name method of the `@nameable`.

#### Prepare CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: decorator.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require_relative 'nameable'
+
```
Comment (MahmoudBakr23) [inline decorator.rb:2]: - [ ] Please add the correct association between the nameable class and the decorator class using the `attr_accessor` 🙏

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
Hello @habasm😃, 

Great work and effort you put into this project so far.

## Highlights 🚀 

There are some excellent points that you have made perfectly. Keep up! 💪
- Good job creating the Namebale and the Decorator classes so far
- Well done working with the correct_name methods in the classes
- Descriptive PR and README file
- Correct Linters

However, there is always room for improvement.

> I'd like to remind you that your long hours of work are appreciated. However, the code review process is for improving your code, not criticizing it. 😄

 ## Required Changes ♻️
_Please check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

N/A

Cheers, and Happy coding!👏👨‍💻💻

Feel free to mention @MahmoudBakr23 to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: decorator.rb (hunk 1 above)
Comment (habasm) [reply decorator.rb:2]: > * [ ]  Please add the correct association between the nameable class and the decorator class using the `attr_accessor` 🙏

@MahmoudBakr23  I am not understood, could you brief it?

Comment (habasm) [reply decorator.rb]: > * [ ]  Please add the correct association between the nameable class and the decorator class using the `attr_accessor` 🙏

I am working on decorator class, and I tried to put based on the solution.

Comment (habasm) [reply decorator.rb]: @MahmoudBakr23 fixed.

File: person.rb [hunk 2]
```diff
@@ -1,26 +1,85 @@
...
     @age >= 18
   end
 end
+
+class BaseDecorator < Nameable
+  def initialize(nameable)
+    super
+
+    @nameable = nameable
+  end
```
Comment (omarsalem7) [inline person.rb:52]: - [ ] While you made the changes requested by the previous reviewer, You removed the `correct_name` method in BaseDecorator, so kindly add it again

File: person.rb [hunk 3]
```diff
@@ -1,26 +1,85 @@
...
+  attr_accessor :name
+
+  def initialize(name)
+    @name = name
+  end
+
+  def correct_name
+    raise NotImplementedError, 'Kindly Implement correct_name method to return the correct name'
+  end
+end
```
Comment (omarsalem7) [inline person.rb:11]: - [ ] While you make changes you put all classes in one file and this is not a best practice, so separate them again and also for `BaseDecorator`, `TrimmerDecorator` , and `CapitalizeDecorator`

**Review by omarsalem7 — CHANGES_REQUESTED** [review]
Hi @habasm 😊, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

Almost near

 ## Highlights
✔️ The Pull Request has a proper title and description
✔️ Correct git-flow

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
- N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb (hunk 3 above)
Comment (habasm) [reply person.rb:11]: @omarsalem7 You Guys I don't know the reason for different review, one said other another said another thing. 😠

Comment (habasm) [reply person.rb]: @omarsalem7 I tried to fix it based on yours also.

**Comment by arslanbisharat** [comment]
Hi @habasm,
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- The Pull Request has a proper title and description. ✔️ 
- You have added a descriptive Readme file. Good Job. ✔️ 
- There is no linter error is present. ✔️ 
- Project has met all the technical aspects you needed for this project. Keep rocking! ✔️ 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**